cogs/Moderation.py

```diff
@@ -39,6 +135,54 @@ async def mkick(
             description=f"Successfully kicked {len(users)} users.",
         ).build()
         await ctx.respond(embed=embed)
+    """
+
+    # Records a MessageFingerprint and returns a matching fingerprint, if there is one.
+    # If there is a matching fingerprint, it is safe to assume that the message is a multipost.
+    async def record_fingerprint(self, message: discord.Message) -> MessageFingerprint | None:
+        fingerprint = MessageFingerprint.build(message)
+
+        # check if any of the fingerprints match
+        matching_fingerprint = None
+        for other_fingerprint in self.fingerprints:
+            if await fingerprint.matches(other_fingerprint):
+                matching_fingerprint = other_fingerprint
+                break
+
+        # this has to happen _after_ the `matching_fingerprint` loop because a fingerprint always matches itself
+        self.fingerprints.append(fingerprint)
+
+        return matching_fingerprint
+
+    # deletes recorded fingerprints after 60 seconds
+    @tasks.loop(seconds=3)
+    async def delete_old_fingerprints_task(self):
+        # new messages are always appended to the end of the list
+        # so we will only be deleting messages from the front of the list
+        # this algorithm finds the number of messages the delete, then deletes them in bulk
+        n_fingerprints_to_delete = 0
+        for fingerprint in self.fingerprints:
+            if time.time() - fingerprint.created_at > 60:
+                n_fingerprints_to_delete += 1
+            else:
+                break
+
+        del self.fingerprints[:n_fingerprints_to_delete]
+
+    @commands.Cog.listener()
+    async def on_message(self, message: discord.Message) -> None:
+        # Bots are allowed to multipost - don't bother fingerprinting their messages
+        if not message.author.bot:
+            previous_message = await self.record_fingerprint(message)
+            if previous_message:
```

Comment:
I'm thinking that maybe DMs should also be excluded

Reply:
DM's must be excluded, trying to delete a message in a DM results in an error

Reply:
Ah right, thanks, updated!